File: app/src/main/java/com/OdiousPanda/thefweather/Utilities/PreferencesUtil.java
package com.OdiousPanda.thefweather.Utilities;

import android.content.Context;
import android.content.SharedPreferences;

public class PreferencesUtil {
    public static final String TEMPERATURE_UNIT = "tempUnit";
    public static final String DISTANCE_UNIT = "distanceUnit";
    public static final String SPEED_UNIT = "speedUnit";
    public static final String PRESSURE_UNIT = "pressureUnit";
    public static final String EXPLICIT_SETTING = "explicitOrNot";
    private static final String BACKGROUND_SETTING = "backgroundType";
    public static final String BACKGROUND_COLOR = "color";
    public static final String BACKGROUND_PICTURE = "picture";
    public static final String BACKGROUND_PICTURE_RANDOM = "picture_random";
    // Shared preferences file name
    private static final String PREF_NAME = "RainbowPreferencesName";
    private static final String IS_NOT_FIRST_TIME_LAUNCH = "IsAppFirstTimeLaunch";
    private static final String WIDGET_ACTION_TAP = "widgetTap";
    private static final String DEFAULT_TEMPERATURE_UNIT = "°C";
    private static final String DEFAULT_DISTANCE_UNIT = "km";
    private static final String DEFAULT_SPEED_UNIT = "kmph";
    private static final String DEFAULT_PRESSURE_UNIT = "psi";
    private static final String DEFAULT_BACKGROUND_SETTING = BACKGROUND_COLOR;
    private static final boolean DEFAULT_EXPLICIT_SETTING = true;
    private static final boolean DEFAULT_NOT_FIRST_TIME_LAUNCH_VALUE = false;

    public static String getBackgroundSetting(Context context){
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getString(BACKGROUND_SETTING, DEFAULT_BACKGROUND_SETTING);
    }

    public static void setBackgroundSetting(Context context, String value){
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putString(BACKGROUND_SETTING, value);
        editor.apply();
    }

    public static boolean isNotFirstTimeLaunch(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getBoolean(IS_NOT_FIRST_TIME_LAUNCH, DEFAULT_NOT_FIRST_TIME_LAUNCH_VALUE);
    }

    public static void setNotFirstTimeLaunch(Context context, boolean notFirstTime) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putBoolean(IS_NOT_FIRST_TIME_LAUNCH, notFirstTime);
        editor.apply();
    }

    public static String getTemperatureUnit(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getString(TEMPERATURE_UNIT, DEFAULT_TEMPERATURE_UNIT);
    }

    public static String getDistanceUnit(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getString(DISTANCE_UNIT, DEFAULT_DISTANCE_UNIT);
    }

    public static String getSpeedUnit(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getString(SPEED_UNIT, DEFAULT_SPEED_UNIT);
    }

    public static String getPressureUnit(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getString(PRESSURE_UNIT, DEFAULT_PRESSURE_UNIT);
    }

    public static boolean isExplicit(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getBoolean(EXPLICIT_SETTING, DEFAULT_EXPLICIT_SETTING);
    }

    public static void setUnitSetting(Context context, String preferenceName, String value) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putString(preferenceName, value);
        editor.apply();
    }

    public static void setExplicitSetting(Context context, boolean value) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putBoolean(EXPLICIT_SETTING, value);
        editor.apply();
    }

    public static int getWidgetTapCount(Context context) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        return pref.getInt(WIDGET_ACTION_TAP, 0);
    }

    public static void setWidgetTapCount(Context context, int count) {
        SharedPreferences pref = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = pref.edit();
        editor.putInt(WIDGET_ACTION_TAP, count);
        editor.apply();
    }
}

File: app/src/main/java/com/odiousPanda/rainbowKt/utilities/PreferencesUtil.kt
package com.odiousPanda.rainbowKt.utilities

import android.content.Context
import com.google.gson.Gson
import com.odiousPanda.rainbowKt.model.dataSource.Coordinate
import java.util.*

object PreferencesUtil {
    const val TEMPERATURE_UNIT = "tempUnit"
    const val DISTANCE_UNIT = "distanceUnit"
    const val SPEED_UNIT = "speedUnit"
    const val PRESSURE_UNIT = "pressureUnit"
    private const val EXPLICIT_SETTING = "explicitOrNot"
    const val BACKGROUND_COLOR = "color"
    const val BACKGROUND_PICTURE = "picture"
    const val BACKGROUND_PICTURE_RANDOM = "picture_random"
    const val NOTIFICATION_SETTING_ON = "on"
    const val NOTIFICATION_SETTING_OFF = "off"
    private const val BACKGROUND_SETTING = "backgroundType"
    private const val LANGUAGE_SETTING = "language"
    private const val CURRENT_LOCATION = "currentLocation"

    // Shared preferences file name
    private const val PREF_NAME = "RainbowPreferencesName"
    private const val WIDGET_ACTION_TAP = "widgetTap"
    private const val DEFAULT_TEMPERATURE_UNIT = "°C"
    private const val DEFAULT_DISTANCE_UNIT = "km"
    private const val DEFAULT_SPEED_UNIT = "kmph"
    private const val DEFAULT_PRESSURE_UNIT = "psi"
    private const val DEFAULT_BACKGROUND_SETTING = BACKGROUND_COLOR
    private const val DEFAULT_EXPLICIT_SETTING = true
    private const val NOTIFICATION_SETTING = "notification"
    private const val NOTIFICATION_TIME = "notificationTime"
    private const val NOTIFICATION_TIME_DEFAULT = "8:00"
    private const val APP_OPEN_COUNT = "App_open_count"
    private const val DEFAULT_LANGUAGE = "en"

    fun getCurrentLocation(context: Context): Coordinate {
        val locationJson = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(CURRENT_LOCATION, "")
        return Gson().fromJson(locationJson, Coordinate::class.java) ?: Coordinate()
    }

    fun setCurrentLocation(context: Context, coordinate: Coordinate) {
        val locationJson = Gson().toJson(coordinate)
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putString(CURRENT_LOCATION, locationJson).apply()
    }

    fun getAppOpenCount(context: Context): Int {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getInt(APP_OPEN_COUNT, 0)
    }

    fun increaseAppOpenCount(context: Context) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putInt(APP_OPEN_COUNT, getAppOpenCount(context) + 1).apply()
    }

    fun getNotificationTime(context: Context): Calendar {
        val calendar = Calendar.getInstance().also { it.timeInMillis = System.currentTimeMillis() }
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(NOTIFICATION_TIME, NOTIFICATION_TIME_DEFAULT)?.split(":")?.toTypedArray()
            ?.let {
                calendar[Calendar.HOUR_OF_DAY] = it[0].toInt()
                calendar[Calendar.MINUTE] = it[1].toInt()
            }
        return calendar
    }

    fun setNotificationTime(context: Context, hour: Int, minute: Int) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putString(NOTIFICATION_TIME, "$hour:$minute").apply()
    }

    fun getNotificationSetting(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(NOTIFICATION_SETTING, NOTIFICATION_SETTING_ON)
    }

    fun setNotificationSetting(context: Context, value: String?) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putString(NOTIFICATION_SETTING, value).apply()

    }

    fun getBackgroundSetting(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(BACKGROUND_SETTING, DEFAULT_BACKGROUND_SETTING)
    }

    fun setBackgroundSetting(context: Context, value: String?) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putString(BACKGROUND_SETTING, value).apply()
    }

    fun getTemperatureUnit(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(TEMPERATURE_UNIT, DEFAULT_TEMPERATURE_UNIT)
    }

    fun getDistanceUnit(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(DISTANCE_UNIT, DEFAULT_DISTANCE_UNIT)
    }

    fun getSpeedUnit(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(SPEED_UNIT, DEFAULT_SPEED_UNIT)
    }

    fun getPressureUnit(context: Context): String? {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getString(PRESSURE_UNIT, DEFAULT_PRESSURE_UNIT)
    }

    fun isExplicit(context: Context): Boolean {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getBoolean(EXPLICIT_SETTING, DEFAULT_EXPLICIT_SETTING)
    }

    fun setUnitSetting(context: Context, preferenceName: String?, value: String?) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putString(preferenceName, value).apply()
    }

    fun setExplicitSetting(context: Context, value: Boolean) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putBoolean(EXPLICIT_SETTING, value).apply()
    }

    fun getWidgetTapCount(context: Context): Int {
        return context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
            .getInt(WIDGET_ACTION_TAP, 0)
    }

    fun setWidgetTapCount(context: Context, count: Int) {
        context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE).edit()
            .putInt(WIDGET_ACTION_TAP, count).apply()
    }
}